src/main/java/frc/trigon/robot/poseestimation/poseestimator/PoseEstimator.java

```diff
@@ -113,41 +152,140 @@ private void logTargetPath() {
     }
 
-    private List<PoseEstimator6328.VisionObservation> getViableVisionObservations() {
-        List<PoseEstimator6328.VisionObservation> viableVisionObservations = new ArrayList<>();
+    private AprilTagCamera[] getCamerasWithNewResult() {
+        final AprilTagCamera[] camerasWithNewResult = new AprilTagCamera[aprilTagCameras.length];
+        int index = 0;
+
         for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
-            final PoseEstimator6328.VisionObservation visionObservation = getVisionObservation(aprilTagCamera);
-            if (visionObservation != null)
-                viableVisionObservations.add(visionObservation);
+            aprilTagCamera.update();
+            if (aprilTagCamera.hasNewResult()) {
+                camerasWithNewResult[index] = aprilTagCamera;
+                index++;
+            }
         }
-        return viableVisionObservations;
+
+        return Arrays.copyOf(camerasWithNewResult, index);
     }
 
-    private PoseEstimator6328.VisionObservation getVisionObservation(AprilTagCamera aprilTagCamera) {
-        aprilTagCamera.update();
-        if (!aprilTagCamera.hasNewResult())
-            return null;
-        final Pose2d robotPose = aprilTagCamera.getEstimatedRobotPose();
-        if (robotPose == null || robotPose.getTranslation() == null || robotPose.getRotation() == null)
-            return null;
+    private void sortCamerasByLatestResultTimestamp(AprilTagCamera[] aprilTagCameras) {
+        QuickSort.sort(aprilTagCameras, AprilTagCamera::getLatestResultTimestampSeconds);
+    }
+
+    /**
+     * Updates the estimated pose from each camera, each at the timestamp of their latest result.
+     */
+    private void updateEstimatedPoseFromVision(AprilTagCamera[] aprilTagCameras) {
+        for (AprilTagCamera aprilTagCamera : aprilTagCameras) {
+            addVisionObservation(
+                    aprilTagCamera.getEstimatedRobotPose(),
+                    aprilTagCamera.calculateStandardDeviations(),
+                    aprilTagCamera.getLatestResultTimestampSeconds()
+            );
+        }
+    }
 
-        return new PoseEstimator6328.VisionObservation(
-                robotPose,
-                aprilTagCamera.getLatestResultTimestampSeconds(),
-                aprilTagCamera.calculateStandardDeviations()
-        );
+    private void addVisionObservation(Pose2d estimatedPose, StandardDeviations standardDeviations, double timestamp) {
+        if (isTimestampOutOfPreviousHeldPosesRange(timestamp))
+            return;
+
+        final Pose2d estimatedPoseAtTimestamp = getPoseAtTimestamp(timestamp);
+        if (estimatedPoseAtTimestamp == null)
+            return;
+
+        final Pose2d estimatedPoseWithAmbiguityCompensation = calculateEstimatedPoseWithAmbiguityCompensation(estimatedPoseAtTimestamp, estimatedPose, standardDeviations);
+        final Transform2d estimatedPoseAtTimestampToOdometryPose = new Transform2d(estimatedPoseAtTimestamp, odometryPose);
```

Comment:
Still wrong

Reply:
How?

Reply:
More than not making sense, this doesn't match the original code

Reply:
Explain how their code works